Does not build with Clang 7 and Debug on Linux

CAF Version: 0.17.0

#### Settings
OS: Linux
Compiler: Clang 7
libcxx: libstdc++
Build Type: Debug

#### Options
Built as Shared library

It's to build nicely, but when liking CAF an undefined reference error occurs:

```
lib  -Llib -lcaf_io -lcaf_openssl -lcaf_core -lm -lssl -lcrypto -ldl -lpthread -lz 
/usr/bin/ld: libcaf_io.so: undefined reference to `caf::sum_type_token<caf::atom_value, 3>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::downstream_msg::batch, 0>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::uri, 5>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<long, 1>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<double, 2>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<std::chrono::duration<long, std::ratio<1l, 1000000000l> >, 4>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::downstream_msg::close, 1>::pos'
/usr/bin/ld: libcaf_io.so: undefined reference to `caf::sum_type_token<caf::dictionary<caf::config_value>, 8>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::downstream_msg::forced_close, 2>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::ipv6_address, 1>::pos'
/usr/bin/ld: libcaf_io.so: undefined reference to `caf::sum_type_token<std::vector<caf::config_value, std::allocator<caf::config_value> >, 7>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::upstream_msg::drop, 2>::pos'
/usr/bin/ld: libcaf_io.so: undefined reference to `caf::sum_type_token<std::string, 6>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<std::string, 0>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::none_t, 0>::pos'
/usr/bin/ld: libcaf_io.so: undefined reference to `caf::sum_type_token<bool, 1>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::upstream_msg::forced_drop, 3>::pos'
/usr/bin/ld: libcaf_io.so: undefined reference to `caf::sum_type_token<long, 0>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::upstream_msg::ack_batch, 1>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::upstream_msg::ack_open, 0>::pos'
```

It ONLY fails for this specific configuration.

Full log: https://travis-ci.com/bincrafters/conan-caf/jobs/220690467#L3616

Yes, same behavior:

```
-- The C compiler identification is Clang 7.0.0
-- The CXX compiler identification is Clang 7.0.0
-- Check for working C compiler: /usr/bin/clang
-- Check for working C compiler: /usr/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/clang++
-- Check for working CXX compiler: /usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: called by CMake conan helper
-- Conan: Adjusting output directories
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Conan: Checking correct version: 7.0
-- Conan: C++ stdlib: libstdc++
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY


-- Build files have been written to: build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3
/usr/bin/cmake -S/home/conan/project/test_package -Bbuild/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3 --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3/CMakeFiles build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3/CMakeFiles/progress.marks
/usr/bin/make -f CMakeFiles/Makefile2 all
make[1]: Entering directory 'build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3'
/usr/bin/make -f CMakeFiles/test_package.dir/build.make CMakeFiles/test_package.dir/depend
make[2]: Entering directory 'build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3'
cd build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3 && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/conan/project/test_package /home/conan/project/test_package build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3 build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3 build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3/CMakeFiles/test_package.dir/DependInfo.cmake --color=
Scanning dependencies of target test_package
make[2]: Leaving directory 'build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3'
/usr/bin/make -f CMakeFiles/test_package.dir/build.make CMakeFiles/test_package.dir/build
make[2]: Entering directory 'build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3'
[ 50%] Building CXX object CMakeFiles/test_package.dir/test_package.cpp.o
/usr/bin/clang++  -D_GLIBCXX_USE_CXX11_ABI=0 -I/home/conan/.conan/data/caf/0.17.0/bincrafters/testing/package/fb25ee98a95e86e5cb1347a55c5dfe38e86b4352/include -I/home/conan/.conan/data/OpenSSL/1.0.2s/conan/stable/package/90cb27019473e527473a04591528c6cd6de29131/include -I/home/conan/.conan/data/zlib/1.2.11/conan/stable/package/35437353b857392790f3121d268b1db3fb9a09b0/include  -m64 -stdlib=libstdc++ -g    -std=gnu++11 -o CMakeFiles/test_package.dir/test_package.cpp.o -c test_package.cpp
[100%] Linking CXX executable bin/test_package
/usr/bin/cmake -E cmake_link_script CMakeFiles/test_package.dir/link.txt --verbose=1
/usr/bin/clang++    -m64 -stdlib=libstdc++ -g        -rdynamic CMakeFiles/test_package.dir/test_package.cpp.o  -o bin/test_package  -L/home/conan/.conan/data/caf/0.17.0/bincrafters/testing/package/fb25ee98a95e86e5cb1347a55c5dfe38e86b4352/lib  -L/home/conan/.conan/data/OpenSSL/1.0.2s/conan/stable/package/90cb27019473e527473a04591528c6cd6de29131/lib  -L/home/conan/.conan/data/zlib/1.2.11/conan/stable/package/35437353b857392790f3121d268b1db3fb9a09b0/lib -Wl,-rpath,/home/conan/.conan/data/caf/0.17.0/bincrafters/testing/package/fb25ee98a95e86e5cb1347a55c5dfe38e86b4352/lib:/home/conan/.conan/data/OpenSSL/1.0.2s/conan/stable/package/90cb27019473e527473a04591528c6cd6de29131/lib:/home/conan/.conan/data/zlib/1.2.11/conan/stable/package/35437353b857392790f3121d268b1db3fb9a09b0/lib -lcaf_core -lcaf_io -lcaf_openssl -lm -lssl -lcrypto -ldl -lpthread -lz 
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::atom_value, 3>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::downstream_msg::batch, 0>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::uri, 5>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<long, 1>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<double, 2>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<std::chrono::duration<long, std::ratio<1l, 1000000000l> >, 4>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::downstream_msg::close, 1>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::dictionary<caf::config_value>, 8>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::downstream_msg::forced_close, 2>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::ipv6_address, 1>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<std::vector<caf::config_value, std::allocator<caf::config_value> >, 7>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::upstream_msg::drop, 2>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<std::string, 6>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<std::string, 0>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::none_t, 0>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<bool, 1>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::upstream_msg::forced_drop, 3>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<long, 0>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::upstream_msg::ack_batch, 1>::pos'
/usr/bin/ld: libcaf_core.so: undefined reference to `caf::sum_type_token<caf::upstream_msg::ack_open, 0>::pos'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/test_package.dir/build.make:87: bin/test_package] Error 1
make[2]: Leaving directory 'build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3'
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/test_package.dir/all] Error 2
make[1]: Leaving directory 'build/a3990596dbbbcab1be157d39d6e5c2e54e5b2bd3'
make: *** [Makefile:87: all] Error 2
```

I know it's annoying reproduce this kind of situation. We have a Docker image with clang7 + cmake with you want. It could save your time:

    docker pull conanio/clang7
    docker run --rm -ti conanio/clang7

--- a/libcaf_core/caf/config_value.hpp
+++ b/libcaf_core/caf/config_value.hpp
@@ -268,7 +268,7 @@ struct sum_type_access<config_value> {
 
   template <class U, int Pos>
   static bool is(const config_value& x, sum_type_token<U, Pos> token) {
-    return x.get_data().is(token.pos);
+    return x.get_data().is(pos(token));
   }
 
   template <class U>
@@ -278,7 +278,7 @@ struct sum_type_access<config_value> {
 
   template <class U, int Pos>
   static U& get(config_value& x, sum_type_token<U, Pos> token) {
-    return x.get_data().get(token.pos);
+    return x.get_data().get(pos(token));
   }
 
   template <class U>
@@ -288,7 +288,7 @@ struct sum_type_access<config_value> {
 
   template <class U, int Pos>
   static const U& get(const config_value& x, sum_type_token<U, Pos> token) {
-    return x.get_data().get(token.pos);
+    return x.get_data().get(pos(token));
   }
 
   template <class U>

--- a/libcaf_core/caf/default_sum_type_access.hpp
+++ b/libcaf_core/caf/default_sum_type_access.hpp
@@ -39,17 +39,17 @@ struct default_sum_type_access {
 
   template <class U, int Pos>
   static bool is(const T& x, sum_type_token<U, Pos> token) {
-    return x.get_data().is(token.pos);
+    return x.get_data().is(pos(token));
   }
 
   template <class U, int Pos>
   static U& get(T& x, sum_type_token<U, Pos> token) {
-    return x.get_data().get(token.pos);
+    return x.get_data().get(pos(token));
   }
 
   template <class U, int Pos>
   static const U& get(const T& x, sum_type_token<U, Pos> token) {
-    return x.get_data().get(token.pos);
+    return x.get_data().get(pos(token));
   }
 
   template <class U, int Pos>

--- a/libcaf_core/caf/sum_type_token.hpp
+++ b/libcaf_core/caf/sum_type_token.hpp
@@ -23,8 +23,11 @@
 namespace caf {
 
 template <class T, int Pos>
-struct sum_type_token {
-  static constexpr auto pos = std::integral_constant<int, Pos>{};
-};
+struct sum_type_token {};
+
+template <class T, int Pos>
+constexpr std::integral_constant<int, Pos> pos(sum_type_token<T, Pos>) {
+  return {};
+}
 
 } // namespace caf

